skylark_library.bzl

```diff
@@ -0,0 +1,103 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Skylib module containing a library rule for aggregating rules files."""
+
+SkylarkLibraryInfo = provider()
+
+def _skylark_library_impl(ctx):
+  all_files = depset(ctx.files.srcs, order="postorder")
+  for dep in ctx.attr.deps:
+    all_files += dep.files
+  return [
+      # All dependent files should be listed in both `files` and in `runfiles`;
+      # this ensures that a `library()` can be referenced as `data` from a
+      # separate program, or from `tools` of a genrule().
+      DefaultInfo(
+          files = all_files,
+          runfiles = ctx.runfiles(files=list(all_files)),
+      ),
+
+      # We also define our own provider struct, for aggregation and testing.
+      SkylarkLibraryInfo(
+          srcs = ctx.files.srcs,
+          transitive_srcs = all_files,
+      ),
+  ]
+
+skylark_library = rule(
+    implementation = _skylark_library_impl,
+    attrs = {
+        "srcs": attr.label_list(
+            allow_files=[".bzl"],
+        ),
+        "deps": attr.label_list(
+            providers=[
+                [SkylarkLibraryInfo],
+                ["files"],
```

Comment:
The `"files"` provider is a break-glass for the case where we need to include Skylark rules from `baselbuild/bazel`, e.g., https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/hash/hash.bzl, and they wouldn't want to break layering to include from bazel-skylib. Is there a better way to limit these to `.bzl` files, or even better, certain whitelisted targets?

Reply:
Exactly, and you're entirely correct. So, is there a filegroup-provided provider that can apply these filters, or does `attr.label_list`'s `allow_files` suffice here? Does that just filter the files in said group?

Reply:
I'm not aware of anything off the top of my head.

One possibility I can think of is, in the impl function for `skylark_library`, just exclude any files that don't end in `.bzl`. It's still hacky, but it at least prevents the case we're trying to avoid. WDYT?

Reply:
Looks like with `allow_files` and use of an `exports_files` in `tools/build_defs/hash/BUILD` (instead of the existing `:srcs` target) would suffice here, and can remove this break-glass.

Reply:
Removed.